src/serde.rs

```diff
@@ -0,0 +1,94 @@
+extern crate serde;
+
+use std::{cmp, fmt};
+use std::marker::PhantomData;
+use self::serde::{Serialize, Serializer, Deserialize, Deserializer, de};
+use super::{Bytes, BytesMut};
+
+macro_rules! serde_impl {
+    ($ty:ident, $visitor_ty:ident) => (
+        impl Serialize for $ty {
+            #[inline]
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where S: Serializer
+            {
+                serializer.serialize_bytes(&self)
+            }
+        }
+
+        struct $visitor_ty<$ty> {
```

Comment:
The type parameter and PhantomData should not be necessary here. This can be `struct $visitor_ty;`.